src/hotspot/share/runtime/mutex.cpp

```diff
@@ -372,17 +371,56 @@ Mutex* Mutex::get_least_ranked_lock_besides_this(Mutex* locks) {
       res = tmp;
     }
   }
+  assert(res != this, "invariant");
+  return res;
+}
+
+// Tests for rank violations that might indicate exposure to deadlock.
+void Mutex::check_rank(Thread* thread) {
+  assert(this->rank() >= 0, "bad lock rank");
+  Mutex* locks_owned = thread->owned_locks();
+
   if (!SafepointSynchronize::is_at_safepoint()) {
-    // In this case, we expect the held locks to be
-    // in increasing rank order (modulo any native ranks)
-    for (tmp = locks; tmp != NULL; tmp = tmp->next()) {
+    // We expect the locks already acquired to be in increasing rank order,
+    // modulo locks of native rank or acquired in try_lock_without_rank_check()
+    for (Mutex* tmp = locks_owned; tmp != NULL; tmp = tmp->next()) {
       if (tmp->next() != NULL) {
-        assert(tmp->rank() == Mutex::native ||
-               tmp->rank() <= tmp->next()->rank(), "mutex rank anomaly?");
+        assert(tmp->rank() == Mutex::native || tmp->rank() < tmp->next()->rank()
+               || tmp->skip_rank_check(), "mutex rank anomaly?");
       }
     }
   }
-  return res;
+
+  // Locks with rank native or suspend_resume are an exception and are not
+  // subject to the verification rules.
+  bool check_can_be_skipped = this->rank() == Mutex::native || this->rank() == Mutex::suspend_resume
+                              || SafepointSynchronize::is_at_safepoint();
+  if (owned_by_self()) {
+    // wait() case
+    Mutex* least = get_least_ranked_lock_besides_this(locks_owned);
+    // We enforce not holding locks of rank special or lower while waiting.
+    // Also "this" should be the monitor with lowest rank owned by this thread.
+    if (least != NULL && (least->rank() <= special ||
+        (least->rank() <= this->rank() && !check_can_be_skipped))) {
+      assert(false, "Attempting to wait on monitor %s/%d while holding lock %s/%d -- "
+             "possible deadlock. %s", name(), rank(), least->name(), least->rank(),
+             least->rank() <= this->rank() ? "Should wait on the least ranked monitor from "
+             "all owned locks." : "Should not block(wait) while holding a lock of rank special.");
+    }
+  } else {
+    // lock()/lock_without_safepoint_check()/try_lock() case
+    Mutex* least = get_least_ranked_lock(locks_owned);
+    // Deadlock prevention rules require us to acquire Mutexes only in
+    // a global total order. For example, if m1 is the lowest ranked mutex
+    // that the thread holds and m2 is the mutex the thread is trying
+    // to acquire, then deadlock prevention rules require that the rank
+    // of m2 be less than the rank of m1. This prevents circular waits.
+    if (least != NULL && least->rank() <= this->rank() && !check_can_be_skipped) {
+      thread->print_owned_locks();
+      assert(false, "Attempting to acquire lock %s/%d out of order with lock %s/%d -- "
+             "possible deadlock", this->name(), this->rank(),least->name(), least->rank());
```

Comment:
nit - missing blank - s/this->rank(),/this->rank(), /

Reply:
Fixed.